test/unitary/session/Session.cpp

```diff
@@ -0,0 +1,425 @@
+            std::copy_n(message.begin(), *len, input_buffer.begin());
+            return true;
+        }
+        else if(std::string("DeleteNoOk") == ::testing::UnitTest::GetInstance()->current_test_info()->name())
+        {
+            return false;
+        }
+        else if(std::string("Listen") == ::testing::UnitTest::GetInstance()->current_test_info()->name())
+        {
+            *len = 0u;
+            *buf = NULL;
+            return true;
+        }
+        else if(std::string("ListenTimeout") == ::testing::UnitTest::GetInstance()->current_test_info()->name())
+        {
+            return false;
+        }
+        else if(std::string("ListenReliably") == ::testing::UnitTest::GetInstance()->current_test_info()->name())
+        {
+            *len = 0u;
+            *buf = NULL;
+            return true;
+        }
+        else if(std::string("ListenReliablyTimeout") == ::testing::UnitTest::GetInstance()->current_test_info()->name())
+        {
+            return false;
+        }
+        else if(std::string("WaitSessionStatusBad") == ::testing::UnitTest::GetInstance()->current_test_info()->name())
+        {
+            SessionTest::listening_counter++;
+            return false;
+        }
+        else if(std::string("RecvMessageOk") == ::testing::UnitTest::GetInstance()->current_test_info()->name())
+        {
+            *len = 8;
+            return true;
+        }
+        else if(std::string("RecvMessageError") == ::testing::UnitTest::GetInstance()->current_test_info()->name())
+        {
+            return false;
+        }
+        return false;
+    }
+
+    static uint8_t comm_error(void)
+    {
+        return 0u;
+    }
+
+    static void on_status_func (struct uxrSession* session, uxrObjectId object_id, uint16_t request_id,
+                             uint8_t status, void* args)
+    {
+        (void) session; (void) object_id; (void) request_id; (void) status; (void) args;
+        if(std::string("ProcessStatus") == ::testing::UnitTest::GetInstance()->current_test_info()->name())
+        {
+            EXPECT_EQ(&SessionTest::current->session, session);
+            EXPECT_EQ(2, object_id.id);
+            EXPECT_EQ(2, object_id.type);
+            EXPECT_EQ(4, request_id);
+            EXPECT_EQ(UXR_STATUS_OK, status);
+            EXPECT_EQ(&SessionTest::current->session, args);
+        }
+    }
+
+    static void on_topic_func (struct uxrSession* session, uxrObjectId object_id, uint16_t request_id,
+                             uxrStreamId stream_id, struct ucdrBuffer* ub, void* args)
+    {
+        (void) session; (void) object_id; (void) request_id; (void) stream_id; (void) ub; (void) args;
+    }
+};
+
+SessionTest* SessionTest::current = nullptr;
+int SessionTest::listening_counter;
+
+TEST_F(SessionTest, SetStatusCallback)
+{
+    int user_data;
+    uxr_set_status_callback(&session, on_status_func, &user_data);
+    //EXPECT_EQ(session.on_status, on_status_func);
```

Comment:
I want to remember that Windows gives me an error because the `==` operator for functions was not given and if I cast it to `void*` it complains too because the invalid cast (even with reinterpret). So... I have not a solution to fix this now.